### PR TITLE
pkp/pkp-lib#6983 On password change, invalidate other sessions

### PR DESCRIPTION
**Pull request by touhidurabir**
This RP aims to fix issue https://github.com/pkp/pkp-lib/issues/6983 . By this enhancement , users will be force logged out from any other devices except the current one when update the password successfully . This takes into the consideration of cases as

1. User password update .
2. User password update by admin/journal manager from journal settings
3. user password reset
4. user password force update on login as password updated or reset

### CODE REVIEW

File: classes/session/SessionDAO.inc.php [hunk 1]
```diff
@@ -192,6 +193,22 @@ public function sessionExistsById($sessionId)
         $row = $result->current();
         return $row ? (bool) $row->row_count : false;
     }
+
+	/**
+     * Delete given user's all sessions or except for the given session id
+     *
+     * @param int                   $userId
```
Comment (asmecher) [inline classes/session/SessionDAO.inc.php:200]: If the annotations are only reiterating the type hints, better to leave them out. If you want to include parameter descriptions (and I recommend it), unfortunately it's currently necessary to restate the type. (Ref for `@param` PHPDoc tag: https://github.com/php-fig/fig-standards/blob/master/proposed/phpdoc-tags.md#510-param)

File: classes/session/SessionManager.inc.php [hunk 2]
```diff
@@ -140,6 +142,31 @@ public static function getManager(): SessionManager
...
+     */
+    public function invalidateSessions(int $userId, string $excludableSessionId = null): bool 
+	{
+        $this->getSessionDao()->deleteUserSessions($userId, $excludableSessionId);
+
+        return true;
+    }
+
+	/**
+     * Get the Session DAO instance assocaited with the current request
```
Comment (asmecher) [inline classes/session/SessionManager.inc.php:161]: *associated

File: controllers/grid/settings/user/form/UserDetailsForm.inc.php [hunk 3]
```diff
@@ -319,6 +320,14 @@ public function execute(...$functionParams)
                 } else {
                     $this->user->setPassword(Validation::encryptCredentials($this->user->getUsername(), $this->getData('password')));
                 }
+
+				$sessionManager = SessionManager::getManager();
```
Comment (asmecher) [inline controllers/grid/settings/user/form/UserDetailsForm.inc.php:324]: Looks like you've got a mix of tabs and spaces here; for `main`, indenting should be 4 spaces. (The auto-formatter should be taking care of this; maybe Captain Hook is not working to set up your git hooks?)

**Review by asmecher — CHANGES_REQUESTED** [review]
Just a couple of small comments, @touhidurabir, thanks!

File: controllers/tab/user/ProfileTabHandler.inc.php [hunk 4]
```diff
@@ -300,15 +301,25 @@ public function changePassword($args, $request)
...
+
+			$sessionManager = SessionManager::getManager();
+            $sessionManager->invalidateSessions($user->getId(), $sessionManager->getUserSession()->getId());
+
             $notificationMgr = new NotificationManager();
-            $notificationMgr->createTrivialNotification($request->getUser()->getId());
-            return new JSONMessage(true);
+            $notificationMgr->createTrivialNotification($user->getId());
+            
+			return new JSONMessage(true);
```
Comment (asmecher) [inline controllers/tab/user/ProfileTabHandler.inc.php:320]: There are still some tab-based indents here; there should be pre-commit git hooks that run the code formatter over the code (using http://captainhook.info). Could you check into whether that's working?

Comment (touhidurabir) [reply controllers/tab/user/ProfileTabHandler.inc.php]: fixed it for this file and other files . should be fine now .

**Pull request merged**